groups/service.go

```diff
@@ -0,0 +1,415 @@
+					}
+					log.Printf("Updated %s to %q as a %s\n", memberEmailId, as.group.EmailId, role)
+				} else {
+					log.Printf("dry-run: would update %s to %q as %s\n", memberEmailId, as.group.EmailId, role)
+				}
+			}
+			continue
+		}
+		member = &admin.Member{
+			Email: memberEmailId,
+			Role:  role,
+		}
+
+		// We did not find the person in the google group, so we add them
+		if config.ConfirmChanges {
+			_, err := as.client.InsertMember(as.group.EmailId, member)
+			if err != nil {
+				return fmt.Errorf("unable to add %s to %q as %s : %v", memberEmailId, as.group.EmailId, role, err)
+			}
+			log.Printf("Added %s to %q as a %s\n", memberEmailId, as.group.EmailId, role)
+		} else {
+			log.Printf("dry-run: would add %s to %q as %s\n", memberEmailId, as.group.EmailId, role)
+		}
+	}
+
+	return nil
+}
+
+func (as *adminService) CreateOrUpdateGroupIfNescessary() error {
+	if *verbose {
+		log.Printf("createOrUpdateGroupIfNecessary %q", as.group.EmailId)
+	}
+
+	group, err := as.client.GetGroup(as.group.EmailId)
+	if err != nil {
+		if apierr, ok := err.(*googleapi.Error); ok && apierr.Code == http.StatusNotFound {
+			if !config.ConfirmChanges {
+				log.Printf("dry-run: would create group %q\n", as.group.EmailId)
+			} else {
+				log.Printf("Trying to create group: %q\n", as.group.EmailId)
+				g := admin.Group{
+					Email: as.group.EmailId,
+				}
+				if as.group.Name != "" {
+					g.Name = as.group.Name
+				}
+				if as.group.Description != "" {
+					g.Description = as.group.Description
+				}
+				g4, err := as.client.InsertGroup(&g)
+				if err != nil {
+					return fmt.Errorf("unable to add new group %q: %v", as.group.EmailId, err)
+				}
+				log.Printf("> Successfully created group %s\n", g4.Email)
+			}
+		} else {
+			return fmt.Errorf("unable to fetch group %q: %#v", as.group.EmailId, err.Error())
+		}
+	} else {
+		if as.group.Name != "" && group.Name != as.group.Name ||
+			as.group.Description != "" && group.Description != as.group.Description {
+			if !config.ConfirmChanges {
+				log.Printf("dry-run: would update group name/description %q\n", as.group.EmailId)
+			} else {
+				log.Printf("Trying to update group: %q\n", as.group.EmailId)
+				g := admin.Group{
+					Email: as.group.EmailId,
+				}
+				if as.group.Name != "" {
+					g.Name = as.group.Name
+				}
+				if as.group.Description != "" {
+					g.Description = as.group.Description
+				}
+				g4, err := as.client.UpdateGroup(as.group.EmailId, &g)
+				if err != nil {
+					return fmt.Errorf("unable to update group %q: %v", as.group.EmailId, err)
+				}
+				log.Printf("> Successfully updated group %s\n", g4.Email)
+			}
+		}
+	}
+	return nil
+}
+
+func (as *adminService) DeleteGroupsIfNecessary() error {
+	g, err := as.client.ListGroups()
+	if err != nil {
+		return fmt.Errorf("unable to retrieve users in domain: %v", err)
+	}
+	if len(g.Groups) == 0 {
+		log.Println("No groups found.")
+		return nil
+	}
```

Comment:
For followup since this is a straight refactor: we don't actually need this, since we'd skip by iterating over nothing below (and I don't think there's value in this logging statement)